lib/realtime/broadcast_changes/handler.ex

```diff
@@ -0,0 +1,205 @@
+  alias Realtime.Adapters.Postgres.Protocol.Write
+  alias Realtime.Api.Tenant
+  alias Realtime.BroadcastChanges.PostgresReplication
+  alias Realtime.Database
+  alias Realtime.Tenants.BatchBroadcast
+  alias Realtime.Tenants.Cache
+
+  defstruct [:tenant_id, relations: %{}, buffer: [], postgres_replication_pid: nil]
+
+  @behaviour PostgresReplication.Handler
+  @registry Realtime.BroadcastChanges.Handler.Registry
+  def start_link(opts), do: GenServer.start_link(__MODULE__, opts, opts)
+
+  @impl true
+  def init(opts) do
+    tenant_id = Keyword.fetch!(opts, :tenant_id)
+
+    tenant = Cache.get_tenant_by_external_id(tenant_id)
+    connection_opts = Database.from_tenant(tenant, "realtime_broadcast_changes", :stop, true)
+
+    supervisor =
+      {:via, PartitionSupervisor,
+       {Realtime.BroadcastChanges.Listener.DynamicSupervisor, tenant_id}}
+
+    name = {:via, Registry, {Realtime.BroadcastChanges.Listener.Registry, tenant_id}}
+
+    configuration = %PostgresReplication{
+      connection_opts: [
+        hostname: connection_opts.host,
+        username: connection_opts.user,
+        password: connection_opts.pass,
+        database: connection_opts.name,
+        port: connection_opts.port,
+        parameters: [
+          application_name: connection_opts.application_name
+        ]
+      ],
+      table: "messages",
+      schema: "realtime",
+      handler_module: __MODULE__,
+      opts: [name: name],
+      metadata: %{tenant_id: tenant_id}
+    }
+
+    children_spec = %{
+      id: Handler,
+      start: {PostgresReplication, :start_link, [configuration]},
+      type: :worker
+    }
+
+    state = %__MODULE__{tenant_id: tenant_id, buffer: [], relations: %{}}
+
+    case DynamicSupervisor.start_child(supervisor, children_spec) do
+      {:ok, pid} ->
+        {:ok, %{state | postgres_replication_pid: pid}}
+
+      {:error, {:already_started, pid}} ->
+        {:ok, %{state | postgres_replication_pid: pid}}
+
+      error ->
+        log_error("UnableToStartPostgresReplication", error)
+        {:stop, error}
+    end
+  end
+
+  @spec name(Tenant.t()) :: term()
+  def name(%Tenant{external_id: tenant_id}) do
+    {:via, Registry, {@registry, tenant_id}}
+  end
+
+  @spec supervisor_spec(Tenant.t()) :: term()
+  def supervisor_spec(%Tenant{external_id: tenant_id}) do
+    {:via, PartitionSupervisor, {Realtime.BroadcastChanges.Handler.DynamicSupervisor, tenant_id}}
+  end
+
+  @impl true
+  def call(message, metadata) when is_write(message) do
+    %{tenant_id: tenant_id} = metadata
+    %Write{message: message} = parse(message)
+
+    case Registry.lookup(@registry, tenant_id) do
+      [{pid, _}] ->
+        message |> decode_message() |> then(&send(pid, &1))
+        :noreply
+
+      _ ->
+        Logger.error("Unable to find BroadcastChanges for tenant: #{tenant_id}")
+        :shutdown
+    end
+  end
+
+  def call(message, _metadata) when is_keep_alive(message) do
```

Comment:
I would consider moving this to PostgrexReplication, as I can't see a handler wanting to do anything but returning the keep alive, and this can be more focused on the specific application aspects.

Reply:
good call will move